File: 4_1/p037.py
def cross(x1, y1, x2, y2):
    return x1*y2 - x2*y1

ax, ay = map(int, input().split())
bx, by = map(int, input().split())
cx, cy = map(int, input().split())
dx, dy = map(int, input().split())

k1 = cross(bx-ax, by-ay, cx-ax, cy-ay)
k2 = cross(bx-ax, by-ay, dx-ax, dy-ay)
k3 = cross(dx-cx, dy-cy, ax-cx, ay-cy)
k4 = cross(dx-cx, dy-cy, bx-cx, by-cy)

if not k1 and not k2 and not k3 and not k4:
    p1 = (ax, ay)
    p2 = (bx, by)
    p3 = (cx, cy)
    p4 = (dx, dy)
    if p1 > p2:
        p1, p2 = p2, p1
    if p3 > p4:
        p3, p4 = p4, p3
    if max(p1, p3) <= min(p2, p4):
        print('Yes')
    else:
        print('No')
else:
    res1 = res2 = False
    if k1 >= 0 and k2 <= 0:
        res1 = True
    elif k1 <= 0 and k2 >= 0:
        res1 = True
    if k3 >= 0 and k4 <= 0:
        res2 = True
    elif k3 <= 0 and k4 >= 0:
        res2 = True
    
    if res1 and res2:
        print('Yes')
    else:
        print('No')

File: 4_1/p037.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 100000000
)

type Point struct {
	X, Y int
}

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
}

func main() {
	scan.Split(bufio.ScanWords)

	p1 := scanPoint()
	p2 := scanPoint()

	p3 := scanPoint()
	p4 := scanPoint()

	c1 := cross(p1, p2, p3)
	c2 := cross(p1, p2, p4)
	c3 := cross(p3, p4, p1)
	c4 := cross(p3, p4, p2)

	if c1 > c2{
		c1, c2 = c2, c1
	}
	if c3 > c4{
		c3, c4 = c4, c3
	}

	if c1 == 0 && c2 == 0 && c3 == 0 && c4 == 0 {
		if compare(p1, p2, p3, p4) {
			fmt.Println("Yes")
		}else{
			fmt.Println("No")
		}
	}else {
		if c1 <= 0 && c2 >= 0 && c3 <= 0 && c4 >= 0{
			fmt.Println("Yes")
		}else{
			fmt.Println("No")
		}
	}
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

func cross(p1, p2, p3 *Point) int {
	return (p2.X-p1.X)*(p3.Y-p1.Y) - (p2.Y-p1.Y)*(p3.X-p1.X)
}

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a >= b {
		return a
	}
	return b
}

func compare(p1, p2, p3, p4 *Point) bool {
	if p1.X == p2.X && p1.X == p3.X && p1.X == p4.X {
		if p1.Y > p2.Y {
			p1, p2 = p2, p1
		}
		if p3.Y > p4.Y{
			p3, p4 = p4, p3
		}
		return max(p1.Y, p3.Y) <= min(p2.Y, p4.Y)
	}else{
		if p1.X > p2.X{
			p1, p2 = p2, p1
		}
		if p3.X > p4.X{
			p3, p4 = p4, p3
		}
		return max(p1.X, p3.X) <= min(p2.X, p4.X)
	}
}

func scanPoint() *Point {
	x := NumStdin()
	y := NumStdin()
	return &Point{x, y}
}
